Spigot-Server-Patches/0594-Damage-with-any-cause.patch

```diff
@@ -0,0 +1,528 @@
++import org.bukkit.event.entity.EntityDamageEvent;
+ import org.bukkit.event.entity.EntityPotionEffectEvent;
+ import org.bukkit.event.player.PlayerTeleportEvent;
+ import org.bukkit.inventory.EntityEquipment;
+@@ -321,7 +324,9 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
+ 
+     @Override
+     public void damage(double amount) {
+-        damage(amount, null);
++        // Paper start - the typecast
++        damage(amount, (org.bukkit.entity.Entity) null);
++        // Paper end
+     }
+ 
+     @Override
+@@ -337,6 +342,24 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
+         entity.damageEntity(reason, (float) amount);
+     }
+ 
++    // Paper start
++
++    @Override
++    public void damage(double amount, EntityDamageEvent.DamageCause damageCause) {
++        DamageSource reason = DamageUtils.damageCauseToDamageSource(damageCause, null);
++        entity.damageEntity(reason, (float) amount);
++    }
++
++    @Override
++    public void damage(double amount, EntityDamageEvent.DamageCause damageCause, org.bukkit.entity.Entity source) {
++        net.minecraft.server.Entity handle = ((CraftEntity) source).getHandle();
++        DamageSource reason = DamageUtils.damageCauseToDamageSource(damageCause, handle);
++        CraftEventFactory.entityDamage = handle;
++        entity.damageEntity(reason, (float) amount);
++    }
++
++    // Paper end
++
+     @Override
+     public Location getEyeLocation() {
+         Location loc = getLocation();
+diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+index 0f0d70c0622558740086f0f7cf89e5500128f955..4d584c423835e2e2d8b038676904bfed30571671 100644
+--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
++++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+@@ -99,6 +99,7 @@ import org.bukkit.craftbukkit.potion.CraftPotionUtil;
+ import org.bukkit.craftbukkit.util.CraftDamageSource;
+ import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+ import org.bukkit.craftbukkit.util.CraftNamespacedKey;
++import org.bukkit.craftbukkit.util.DamageUtils;
+ import org.bukkit.entity.AbstractHorse;
+ import org.bukkit.entity.Animals;
+ import org.bukkit.entity.AreaEffectCloud;
+@@ -886,10 +887,16 @@ public class CraftEventFactory {
+     }
+ 
+     private static EntityDamageEvent handleEntityDamageEvent(Entity entity, DamageSource source, Map<DamageModifier, Double> modifiers, Map<DamageModifier, Function<? super Double, Double>> modifierFunctions, boolean cancelled) {
++        // Paper start
++        Entity damager = entityDamage;
++        entityDamage = null;
++        // Paper end
+         if (source.isExplosion()) {
+             DamageCause damageCause;
+-            Entity damager = entityDamage;
+-            entityDamage = null;
++            // Paper start - moved above conditional block
```

Comment:
Is this safe?

Reply:
Yes, thats pretty much the whole point of this patch - support damager entity/source context for all damage causes.

Reply:
But is setting the `entityDamage` field safe when the source *isn't* an explosion?

Reply:
I haven't found any issues while testing.